### DISCUSSION TITLE
How are booleans in Gleam represented?

### DISCUSSION THREAD

**Discussion by CrowdHailer** [gleam-lang/gleam#1026] [Questions & support] [answered]
There are typed expressions for other literals such as Int/Float. But as far as I can see not a similar expression for Booleans

https://github.com/gleam-lang/gleam/blob/87f3bb01e9362106b16b164f133f6b448f2e6762/src/erl.rs#L1420-L1422

How are Booleans represented?

**Comment by lpil** [comment, marked as answer]
Booleans are not a primitive type with a dedicated literal syntax, they're just a normal custom type but defined in the prelude.

```rust
pub type Bool {
  True
  False
}
```

**Reply by CrowdHailer** [reply]
When printing the erlang I guess this works because a type without fields just gets generated as an atom. and the atom true/false is the same as idiomatic erlang booleans.

I think for the JS, I'm going to have to recognise this prelude record and choose to render them as JS bools for interop.

Is the same true for Nil, i.e. it's a single variant record that helpfully gets output as the relevant nil atom.

**Reply by lpil** [reply]
Yup we'd want to render something special for bools and nil, likely `true`, `false` and `undefined`. We can do this by checking the type of the constructor and seeing if it's the prelude `Bool` or `Nil` respectively

**Reply by Rudxain** [reply]
This is genuinely beautiful. I wish Rust `bool`s were just 2-variant `enum`s, although this may have some caveats in that lang (unless fixed by special traits)

I just learned that [Ada does it](https://learn.adacore.com/courses/intro-to-ada/chapters/strongly_typed_language.html#enumerations) too!:
```ada
type Boolean is (False, True);
```
